ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -974,44 +972,40 @@ Pair<Boolean, String> canCBOHandleAst(ASTNode ast, QB qb, PreCboCtx cboCtx) {
    *         Query<br>
    *         2. Nested Subquery will return false for qbToChk.getIsQuery()
    */
-  private static String canHandleQbForCbo(QueryProperties queryProperties, HiveConf conf,
-      boolean topLevelQB, boolean verbose) {
-
-    if (!queryProperties.hasClusterBy() && !queryProperties.hasDistributeBy()
-        && !(queryProperties.hasSortBy() && queryProperties.hasLimit())
-        && !queryProperties.hasPTF() && !queryProperties.usesScript()
-        && queryProperties.isCBOSupportedLateralViews()) {
-      // Ok to run CBO.
-      return null;
-    }
-
+  private static String canHandleQbForCbo(QueryProperties queryProperties,
+                                          HiveConf conf, boolean topLevelQB) {
+    List reasons = new ArrayList();
     // Not ok to run CBO, build error message.
-    String msg = "";
-    if (verbose) {
-      if (queryProperties.hasClusterBy()) {
-        msg += "has cluster by; ";
-      }
-      if (queryProperties.hasDistributeBy()) {
-        msg += "has distribute by; ";
-      }
-      if (queryProperties.hasSortBy() && queryProperties.hasLimit()) {
-        msg += "has sort by with limit; ";
-      }
-      if (queryProperties.hasPTF()) {
-        msg += "has PTF; ";
-      }
-      if (queryProperties.usesScript()) {
-        msg += "uses scripts; ";
-      }
-      if (queryProperties.hasLateralViews()) {
-        msg += "has lateral views; ";
-      }
-      if (msg.isEmpty()) {
-        msg += "has some unspecified limitations; ";
-      }
-      msg = msg.substring(0, msg.length() - 2);
+    String errorMsg = "";
+    if (queryProperties.hasClusterBy()) {
+      errorMsg = "has cluster by";
+      reasons.add(errorMsg);
+    }
+    if (queryProperties.hasDistributeBy()) {
+      errorMsg = "has distribute by";
+      reasons.add(errorMsg);
+    }
+    if (queryProperties.hasSortBy() && queryProperties.hasLimit()) {
+      errorMsg = "has sort by with limit";
+      reasons.add(errorMsg);
+    }
+    if (queryProperties.hasPTF()) {
+      errorMsg = "has PTF";
+      reasons.add(errorMsg);
+    }
+    if (queryProperties.usesScript()) {
+      errorMsg = "uses scripts";
+      reasons.add(errorMsg);
+    }
+    if (queryProperties.hasLateralViews()) {
+      errorMsg = "has lateral views";
+      reasons.add(errorMsg);
+    }
+    if (!queryProperties.isCBOSupportedLateralViews()) {
```

Comment:
Trying to keep the error messaging the same. If isCBOSupportedLateralViews() returns False, only then the error message will be "has some unspecified limitations".